monkey/infection_monkey/telemetry/messengers/batching_telemetry_messenger.py

```diff
@@ -0,0 +1,94 @@
+import queue
+import threading
+import time
+from typing import Dict
+
+from infection_monkey.telemetry.i_batchable_telem import IBatchableTelem
+from infection_monkey.telemetry.i_telem import ITelem
+from infection_monkey.telemetry.messengers.i_telemetry_messenger import ITelemetryMessenger
+
+DEFAULT_PERIOD = 5
+WAKES_PER_PERIOD = 4
+
+
+class BatchingTelemetryMessenger(ITelemetryMessenger):
+    """
+    An ITelemetryMessenger decorator that aggregates IBatchableTelem telemetries
+    and periodically sends them to Monkey Island.
+    """
+
+    def __init__(self, telemetry_messenger: ITelemetryMessenger, period=DEFAULT_PERIOD):
+        self._queue: queue.Queue[ITelem] = queue.Queue()
+        self._thread = BatchingTelemetryMessenger._BatchingTelemetryMessengerThread(
+            self._queue, telemetry_messenger, period
+        )
+
+        self._thread.start()
+
+    def __del__(self):
+        self._thread.stop()
+
+    def send_telemetry(self, telemetry: ITelem):
+        self._queue.put(telemetry)
+
+    class _BatchingTelemetryMessengerThread:
+        def __init__(self, queue: queue.Queue, telemetry_messenger: ITelemetryMessenger, period):
+            self._queue: queue.Queue[ITelem] = queue
+            self._telemetry_messenger = telemetry_messenger
+            self._period = period
+
+            self._should_run_batch_thread = True
```

Comment:
Why is this variable required? It's only used in `_manage_telemetry_batches`, and if `False` only if thread shoud stop. But if the thread should stop, it's joined and `_manage_telemetry_batches` no longer runs.

Reply:
Joining a thread doesn't stop it. It waits for the thread to stop before execution continues. A mechanism is required to signal the thread to stop. In this case, it's the state of `self._should_run_batch_thread`.